semgrep-core/src/engine/Match_patterns.ml

```diff
@@ -270,7 +271,25 @@ let check2 ~hook range_filter config rules equivs (file, lang, ast) =
              Common.push (pattern, strs, rule, cache) rules
            in
            match any with
-           | E pattern -> push_with_annotation any pattern expr_rules
+           | E pattern -> (
+               push_with_annotation any pattern expr_rules;
+               match pattern.e with
+               | Assign (_, tok, _) when config.Config.vardef_assign ->
+                   (* vardef-assign: let `$X = $E` match e.g. `var x = 1;`
+                    *
+                    * old: We used to have this equivalence built into
+                    * Visitor_AST. This was error prone because visitors
+                    * typically do side-effectful things and VarDefs were
+                    * visited twice (as a VarDef and as an Assign).
+                    *
+                    * If we just add an ExprStmt-pattern to stmt_rules,
+                    * the "vardef ==> assign" equivalence in
+                    * Generic_vs_generic.m_stmt will take care of the rest.
```

Comment:
hmm maybe we need to use config.Config.vardef_assign in m_stmts too then?

Reply:
True, that's right!